File: ai-service/app/cache.py
"""Shared cache layer -- Redis-backed when settings.redis_url is configured,
an in-process dict otherwise, so callers work identically in both
environments and nothing crashes if Redis is unset or briefly unreachable.

Mirrors lib/cache.ts's design exactly (same get-or-set contract, same
graceful-degrade behavior) -- see that file's docstring for the full
rationale: a horizontally-scaled deploy has one cache per instance without
this, so an expensive, repeatedly-hit computation (here: /ai/render/{slug}'s
preview.run(), which re-executes every widget's live query) re-pays its full
cost once per instance instead of once total.
"""
from __future__ import annotations

import json
import time
from typing import Any, Awaitable, Callable, TypeVar

from .config import get_settings

T = TypeVar("T")

# `_redis_client` states: unset (None) = not yet attempted OR attempted and
# unavailable -- `_redis_attempted` distinguishes the two so a failed connect
# is not retried on every single cache call.
_redis_client: Any = None
_redis_attempted = False

_memory_store: dict[str, tuple[float, str]] = {}  # key -> (expires_at_monotonic, json_value)


async def _get_redis() -> Any:
    global _redis_client, _redis_attempted
    if _redis_attempted:
        return _redis_client
    _redis_attempted = True
    url = get_settings().redis_url
    if not url:
        return None
    try:
        import redis.asyncio as redis_asyncio  # optional dependency; see requirements.txt
        client = redis_asyncio.from_url(url, decode_responses=True, socket_connect_timeout=2)
        await client.ping()
        _redis_client = client
    except Exception:
        _redis_client = None
    return _redis_client


def _memory_get(key: str) -> str | None:
    hit = _memory_store.get(key)
    if not hit:
        return None
    expires_at, value = hit
    if expires_at < time.monotonic():
        _memory_store.pop(key, None)
        return None
    return value


def _memory_set(key: str, value: str, ttl_seconds: int) -> None:
    _memory_store[key] = (time.monotonic() + ttl_seconds, value)


async def cache_get(key: str) -> Any | None:
    client = await _get_redis()
    if client is not None:
        try:
            raw = await client.get(key)
            return json.loads(raw) if raw is not None else None
        except Exception:
            pass  # transient Redis failure -- fall through to the memory store
    raw = _memory_get(key)
    return json.loads(raw) if raw is not None else None


async def cache_set(key: str, value: Any, ttl_seconds: int) -> None:
    serialized = json.dumps(value)
    client = await _get_redis()
    if client is not None:
        try:
            await client.set(key, serialized, ex=max(1, int(ttl_seconds)))
            return
        except Exception:
            pass  # fall through to memory
    _memory_set(key, serialized, ttl_seconds)


async def get_or_set(key: str, ttl_seconds: int, fn: Callable[[], Awaitable[T]]) -> T:
    """Returns the cached JSON-serializable value for `key` if fresh,
    otherwise calls fn(), caches the result, and returns it."""
    cached = await cache_get(key)
    if cached is not None:
        return cached
    value = await fn()
    await cache_set(key, value, ttl_seconds)
    return value

File: lib/cache.ts
/**
 * Shared cache layer -- Redis-backed when REDIS_URL is configured, an
 * in-process Map otherwise, so every caller works identically in both
 * environments and nothing crashes if Redis is unset or briefly unreachable.
 *
 * WHY THIS EXISTS: lib/lms-learnworlds.ts already had its own ad hoc
 * in-process `_cache` Map (ROADMAP Phase 6's "known target: /api/dashboard/lms
 * cold call measured at 14.2s") -- correct for a single instance, but on a
 * horizontally-scaled deploy (multiple Next.js instances/serverless
 * invocations) each instance has its own empty cache, so the same tenant's
 * expensive LearnWorlds aggregation re-runs once per instance instead of
 * once total. This module lets that kind of cache be shared across
 * instances via Redis, while still degrading gracefully to the old
 * per-instance behavior wherever REDIS_URL isn't set (e.g. local dev).
 *
 * SERVER-ONLY. ioredis is dynamically imported so this file has zero cost
 * for any caller that never configures REDIS_URL.
 */

interface MemoryEntry {
  expiresAt: number
  value: string
}

// `undefined` = connection not yet attempted; `null` = attempted and
// unavailable (unset REDIS_URL, or connect failed) -- distinct states so a
// failed connect attempt is not retried on every single cache call.
let redisClient: import('ioredis').Redis | null | undefined
let redisInitPromise: Promise<import('ioredis').Redis | null> | null = null

const memoryStore = new Map<string, MemoryEntry>()

async function getRedis(): Promise<import('ioredis').Redis | null> {
  if (redisClient !== undefined) return redisClient
  if (redisInitPromise) return redisInitPromise

  redisInitPromise = (async () => {
    const url = process.env.REDIS_URL
    if (!url) {
      redisClient = null
      return null
    }
    try {
      const { default: Redis } = await import('ioredis')
      const client = new Redis(url, {
        lazyConnect: true,
        maxRetriesPerRequest: 1,
        connectTimeout: 2_000,
        // Never let a slow/unreachable Redis hang a request queue forever --
        // one failed retry and callers fall back to the in-process store.
        retryStrategy: () => null,
      })
      client.on('error', () => { /* swallowed -- callers fall back to memory */ })
      await client.connect()
      redisClient = client
      return client
    } catch {
      redisClient = null
      return null
    }
  })()

  return redisInitPromise
}

function memoryGet<T>(key: string): T | null {
  const hit = memoryStore.get(key)
  if (!hit) return null
  if (hit.expiresAt < Date.now()) {
    memoryStore.delete(key)
    return null
  }
  return JSON.parse(hit.value) as T
}

function memorySet(key: string, serialized: string, ttlSeconds: number): void {
  memoryStore.set(key, { expiresAt: Date.now() + ttlSeconds * 1000, value: serialized })
}

export async function cacheGet<T>(key: string): Promise<T | null> {
  const client = await getRedis()
  if (client) {
    try {
      const raw = await client.get(key)
      return raw !== null ? (JSON.parse(raw) as T) : null
    } catch {
      // Redis reachable at connect time but failing now (e.g. transient
      // network blip) -- fall back to whatever the in-process store has
      // rather than treating this request as a hard miss.
    }
  }
  return memoryGet<T>(key)
}

export async function cacheSet(key: string, value: unknown, ttlSeconds: number): Promise<void> {
  const serialized = JSON.stringify(value)
  const client = await getRedis()
  if (client) {
    try {
      await client.set(key, serialized, 'EX', Math.max(1, Math.trunc(ttlSeconds)))
      return
    } catch {
      // fall through to memory
    }
  }
  memorySet(key, serialized, ttlSeconds)
}

/**
 * Returns the cached value for `key` if fresh, otherwise calls fn(), caches
 * the result, and returns it. The one call most callers actually want.
 */
export async function getOrSetCache<T>(key: string, ttlSeconds: number, fn: () => Promise<T>): Promise<T> {
  const cached = await cacheGet<T>(key)
  if (cached !== null) return cached
  const value = await fn()
  await cacheSet(key, value, ttlSeconds)
  return value
}
